### PR TITLE
fix(billing): charge redirected per-request image responses

### PR DESCRIPTION
**Pull request by ding113**
## Summary
Fixes billing for redirected image requests when using per-request pricing (\`input_cost_per_request\`). Previously, successful image API responses without token usage data would bypass billing entirely when the resolved local price contained a per-request fee. This ensures proper cost tracking for OpenAI Images API and similar endpoints that don't return token usage.

## Problem
After the fix in #1118 for image requests without usage payload, a gap remained: when model redirects occur with \`billingModelSource\` set to \`original\` or \`redirected\`, successful responses without token usage (e.g., OpenAI Images API) were not being charged even when the local price table configured \`input_cost_per_request\`.

This caused:
- Missing billing records for redirected image generation requests with per-request pricing
- Zero-cost entries when providers didn't return usage payloads, despite local pricing having per-request fees
- Inconsistent billing behavior between regular chat completions and image API requests

**Related Issues:**
- Follow-up to #1118 - fixes remaining gap for model redirect scenarios with per-request pricing

## Solution
Introduced \`resolveBillableUsageMetricsForCost()\` to synthesize billable usage metrics when:
1. The response is successful (2xx status)
2. No token usage is present in the response
3. The resolved pricing (based on \`billingModelSource\`) contains a valid \`input_cost_per_request\`

This allows the existing billing pipeline (\`updateRequestCostFromUsage\`, \`trackCostToRedis\`) to process the request with the per-request fee from the correct model (either pre-redirect or post-redirect based on settings).

Key changes:
- Added \`hasBillableInputCostPerRequest()\` to validate per-request pricing availability
- Added \`resolveBillableUsageMetricsForCost()\` to generate empty usage objects that trigger per-request billing
- Unified the billing path in both streaming and non-streaming response handling
- Preserved existing model billing-source semantics: \`original\` charges pre-redirect model, \`redirected\` charges post-redirect model

### Core Changes
- \`src/app/v1/_lib/proxy/response-handler.ts\`: Added \`resolveBillableUsageMetricsForCost()\` and integrated it into non-streaming, streaming, and request finalization paths

### Test Coverage
- \`tests/integration/billing-model-source.test.ts\`: Added regression tests for \`/v1/images/edits\` with \`gpt-image-2 -> gpt-image-2-all\` redirect scenarios, validating:
  - Provider multiplier application (2x)
  - Group multiplier application (3x)
  - DB cost recording with proper breakdown
  - Session cost tracking
  - Rate-limit cost tracking

### Automated Tests
- [x] Integration tests added for redirected per-request image billing
- [x] Tests cover both \`original\` and \`redirected\` billing model sources
- [x] Tests verify provider multiplier, group multiplier, and cost breakdown accuracy

### Test Commands
\`\`\`bash
bunx vitest run tests/integration/billing-model-source.test.ts -t \"模型重定向后的图片按次计费\"
bunx vitest run tests/integration/billing-model-source.test.ts
\`\`\`

### Verification
\`\`\`bash
bun run build      # Production build (passes)
bun run lint       # Biome check (passes)
bun run lint:fix   # Auto-fix (passes)
bun run typecheck  # TypeScript check (passes)
bun run test       # All tests (passes)
\`\`\`

## Breaking Changes
None. This is a bug fix that enables previously non-functional billing paths. No existing billing behavior changes - only adds missing billing for the edge case.

## Checklist
- [x] Code follows project conventions
- [x] Self-review completed
- [x] Tests pass locally
- [x] No documentation updates needed (behavioral fix, transparent to users)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes a billing gap introduced after #1118: successful image API responses (e.g., `/v1/images/edits`) that carry no token-usage payload were not being charged when the resolved local pricing contained `input_cost_per_request` and a model redirect was in play. The fix introduces `resolveBillableUsageMetricsForCost()`, which synthesizes an empty `UsageMetrics` (`{}`) that routes through the existing billing pipeline (`updateRequestCostFromUsage`, `trackCostToRedis`, session observability) using the correct pre- or post-redirect model according to `billingModelSource` semantics. Test coverage is comprehensive across both billing model sources, provider/group multipliers, zero-cost guard, pricing-DB failure, fake-200 detection, and `finalizeRequestStats` error-message preservation.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — focused billing bug fix with no regressions to existing token-based billing paths.

No P0/P1 issues found. The new `resolveBillableUsageMetricsForCost` guard correctly short-circuits for non-billing endpoints, non-2xx responses, zero/missing per-request prices, and fake-200 error payloads. The `usageRecord` removal is correct (it was redundant given the `billableUsageMetrics` guard). The session-observability condition change (`usageMetrics || costUsdStr !== undefined`) is an improvement for zero-token responses. All three billing paths (non-stream, SSE, `finalizeRequestStats`) are treated consistently, and the existing pricing cache prevents double resolution. Test coverage is thorough across all edge cases called out in the PR description.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/proxy/response-handler.ts | Adds `resolveBillableUsageMetricsForCost()` and integrates it into non-streaming, SSE, and `finalizeRequestStats` paths; removes the now-redundant `usageRecord` gate; corrects session-observability condition to trigger on `costUsdStr` for zero-token requests; SSE `costUsd` payload correctly split out of the `usageForCost` guard. |
| tests/integration/billing-model-source.test.ts | Adds six focused regression tests covering both `billingModelSource` values, zero per-request price, pricing-DB failure, fake-200 error payload, and `finalizeRequestStats` error-message propagation; extends `createSession` with `requestPath` and `groupCostMultiplier` parameters; adds `makePriceRecord` `source` field. |

</details>

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Response received] --> B{usageMetrics present?}
    B -- Yes --> C[normalizeUsageWithSwap]
    C --> D[resolveBillableUsageMetricsForCost returns usageMetrics]
    B -- No --> E[resolveBillableUsageMetricsForCost]
    E --> F{isNonBillingEndpoint?}
    F -- Yes --> G[return null - skip billing]
    F -- No --> H{statusCode 2xx?}
    H -- No --> G
    H -- Yes --> I[getResolvedPricingByBillingSource]
    I --> J{hasBillableInputCostPerRequest gt 0?}
    J -- No --> G
    J -- Yes --> K{responseText provided?}
    K -- Yes --> L{detectUpstreamError}
    L -- isError=true --> G
    L -- isError=false --> M[return empty UsageMetrics]
    K -- No --> M
    D --> N[updateRequestCostFromUsage + trackCostToRedis + updateSessionUsage]
    M --> N
    G --> O[skip billing]
```
</details>

<sub>Reviews (5): Last reviewed commit: ["fix(billing): skip per-request billing f..."](https://github.com/ding113/claude-code-hub/commit/8c69c56294dc5811868f8e78101877f661465c22) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29830994)</sub>

### CODE REVIEW

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 1]
```diff
@@ -411,6 +411,43 @@ function isNonBillingUsageEndpoint(session: ProxySession): boolean {
...
 }
 
+function hasBillableInputCostPerRequest(priceData: { input_cost_per_request?: unknown }): boolean {
+  const inputCostPerRequest = priceData.input_cost_per_request;
+  return (
+    typeof inputCostPerRequest === "number" &&
+    Number.isFinite(inputCostPerRequest) &&
+    inputCostPerRequest >= 0
+  );
+}
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts:421]: 已在 9d9725d5 修复：`hasBillableInputCostPerRequest` 收紧为 `> 0`，并补了 `/v1/images/edits` + `gpt-image-2 -> gpt-image-2-all` 的 0 元按次回归测试，断言不会调用 `updateMessageRequestCostWithBreakdown`、`SessionManager.updateSessionUsage`、`RateLimitService.trackCost`。

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 2]
```diff
@@ -411,6 +411,43 @@ function isNonBillingUsageEndpoint(session: ProxySession): boolean {
...
+
+  const resolvedPricing = await session.getResolvedPricingByBillingSource(provider);
+  if (!resolvedPricing?.priceData || !hasBillableInputCostPerRequest(resolvedPricing.priceData)) {
+    return null;
+  }
+
+  // 成功响应可能没有 token usage（例如 OpenAI Images），但本地价格表仍可配置按次价格。
+  // 这里用空 usage 只承载 input_cost_per_request，不新增按图、按 token 等语义。
+  return {};
+}
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts:476]: 已核对：`ProxySession.getResolvedPricingByBillingSource` 本身有 request-scoped `resolvedPricingCache`，cache key 包含计费来源、原始模型、重定向模型和 provider identity；同一次请求里后续调用会命中内存 Map，不会重复查 `findLatestPriceByModel`。`trackCostToRedis` 也已经接收 `costUpdateResult.resolvedPricing` override。为了把本 PR 聚焦在“重定向后本地手动按次价格不生效”的回归修复上，这里不扩大改动成 response-handler 成本 API 重构。

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 3]
```diff
@@ -3602,6 +3651,56 @@ export async function finalizeRequestStats(
...
+      session.shouldTrackSessionObservability()
+    ) {
+      void SessionManager.updateSessionUsage(session.sessionId, {
+        costUsd: perRequestCostUsd,
+        status: statusCode >= 200 && statusCode < 300 ? "completed" : "error",
+        statusCode,
+      }).catch((error: unknown) => {
+        logger.error("[ResponseHandler] Failed to update session usage:", error);
+      });
+    }
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts:3733]: 已核对这条重构建议。当前无 usage 分支只在 2xx 且存在正数 `input_cost_per_request` 时进入，且 Redis 成本追踪已经沿用 `trackCostToRedis` 的 fire-and-forget 外层调度语义。把 finalize 的无 usage 分支和常规 token usage 分支进一步合并会触碰正常 token 计费收尾路径，超出本次 bugfix 的必要范围；本 PR 保持局部修复，避免引入无关行为变化。

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 4]
```diff
@@ -411,6 +411,54 @@ function isNonBillingUsageEndpoint(session: ProxySession): boolean {
...
+    return null;
+  }
+
+  if (!resolvedPricing?.priceData || !hasBillableInputCostPerRequest(resolvedPricing.priceData)) {
+    return null;
+  }
+
+  // 成功响应可能没有 token usage（例如 OpenAI Images），但本地价格表仍可配置按次价格。
+  // 这里用空 usage 只承载 input_cost_per_request，不新增按图、按 token 等语义。
+  return {};
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts:475]: 已在 8c69c562 修复：`resolveBillableUsageMetricsForCost` 现在会在合成空 usage 前检查响应体，并用 `detectUpstreamErrorFromSseOrJsonText(responseText, { maxJsonCharsForMessageCheck: 0 })` 只启用强 fake-200 信号。命中顶层 `error` / `error.message` / HTML / empty body 时跳过按次计费，不写 DB cost、session usage 或 rate-limit cost。已补 `/v1/images/edits` + `gpt-image-2 -> gpt-image-2-all` 的回归测试，覆盖 `200 + {"error":{"message":"invalid api key"}}` 仍透传 200 但不计费。

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 5]
```diff
@@ -411,6 +411,43 @@ function isNonBillingUsageEndpoint(session: ProxySession): boolean {
...
+
+  if (usageMetrics) {
+    return usageMetrics;
+  }
+
+  if (statusCode < 200 || statusCode >= 300) {
+    return null;
+  }
+
+  const resolvedPricing = await session.getResolvedPricingByBillingSource(provider);
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts]: 已在 c04b4224 修复并补测试：`resolveBillableUsageMetricsForCost` 对 `getResolvedPricingByBillingSource` 的异常做 catch，价格查询失败时跳过按次计费，不再把成功的上游响应改写成内部失败；回归测试覆盖了 pricing db unavailable 时响应仍为 200 且不写 DB/session/限流成本。

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 6]
```diff
@@ -411,6 +411,43 @@ function isNonBillingUsageEndpoint(session: ProxySession): boolean {
...
```
[the remark this answers is not in the document]
Comment (ding113) [reply src/app/v1/_lib/proxy/response-handler.ts]: 已复核这条性能建议：`ProxySession.getResolvedPricingByBillingSource` 有 request-scoped `resolvedPricingCache`，同一请求内第二次调用会命中内存缓存，不会重复查 `findLatestPriceByModel`。本 PR 保持在计费 bugfix 范围内，没有把 cost API 扩成返回 pricing 的结构重构。
